CloudSqlImportOptimization/README.md

```diff
@@ -0,0 +1,190 @@
+instance.
+
+These optimization include:
+
+*   Primary Key ordered data dumps
+*   Loading schemas and data separately
+*   Dropping secondary indexes for speedier ingestion
+*   Chunking table data into similarly sized chunks for speedier ingestion.
+*   Loading data in parallel in multiple groups
+*   Restoring secondary indexes after data load
+*   Performing secondary indexes restores in parallel
+
+We've also included quality of life features:
+
+*   The MySQL dumper tool to dump your database in an importer compatible format
+*   Checksum checks to verify the data integrity of the import
+*   Checkpoints to resume failed import operations
+*   Optionally dropping databases that failed to be imported
+*   Low performance penalties from running on a local Linux machine vs Google
+    Compute Engine
+
+## Prerequisites
+
+*   If you are running our tools on prem please ensure you have configured
+    CloudSQL with
+    [Public IP Connectivity](https://cloud.google.com/sql/docs/mysql/connect-admin-ip)
+*   If you are running our tools from whithin GCP please ensure you have
+    configured CloudSQL with
+    [Private IP Connectivity](https://cloud.google.com/sql/docs/mysql/private-ip)
+*   For exports ensure your MySQL user has regular admin privileges and
+    **FILE**.
+*   For imports ensure your CloudSQL user has admin privileges including the
+    ability to create, populate and possibly, drop databases.
+*   If you want use SSL/TLS connections please
+    [configure SSL/TLS in CloudSQL.](https://cloud.google.com/sql/docs/mysql/configure-ssl-instance)
+
+## Getting Started
+
+### Exporting a copy of your MySQL database
+
+Before being able to run the importer you'll need to have dumped copy of your
+database. Unfortunately, a monolithic dump file won't work with the importer
+therefore, we provide an exporter tool that can be used to generate dumps in
+correct format. You can run the dumper by calling:
+
+       $ ./mysql_exporter.py <database-name> <destination-folder>
+
+To get a list of supported arguments run:
+
+       $ ./mysql_exporter.py -h
+
+TIP: Optional parameters have sensible defaults in place.
+
+TIP: You can use the **--verbose** options for extra logging.
+
+#### Limitations
+
+WARNING: Please note that the exporter tool **must be run on the same computer
+hosting your MySQL database**. This is because we use **mysqldump's --tab
+option** to dump each table's data into separate files.
+
+TIP: For more information on the limitations of using **--tab** see
+[mysqldump's docs.](https://dev.mysql.com/doc/refman/8.0/en/mysqldump.html#option_mysqldump_tab)
+
+In case runnning the exporter locally is a no-go in your environment here are
+the steps necessary for you to produce a MySQL dump that will be compatible with
+the importer:
+
+*   Get your database's schema by calling `mysqldump --no-data`
+*   Ensure the database schema follows the following naming convention:
+    <database_name>_schema.sql
```

Comment:
`...`

Reply:
The tool needn't run in GCE but I can make a note if need be, thoughts?

Reply:
correct; my point was if you us GCE, you need a good strong instance